llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/TaskExecutorService.java

```diff
@@ -650,9 +647,16 @@ public boolean updateFragment(String fragmentId, boolean isGuaranteed) {
         taskWrapper.updateIsGuaranteed(isGuaranteed);
         forceReinsertIntoQueue(taskWrapper, isRemoved);
       } else {
-        taskWrapper.updateIsGuaranteed(isGuaranteed);
-        if (!isGuaranteed && !taskWrapper.isInPreemptionQueue()) {
-          // No need to check finishable here; if it was set it would already be in the queue.
+        // if the task is guaranteed it cannot be evicted, so remove it from pre-emption queue
+        if (isGuaranteed) {
+          // removals are idempotent, so no other state checks are required
+          removeFromPreemptionQueue(taskWrapper);
+        } else {
```

Comment:
I guess this can be further simplified to removeFromPreemptionQueue and if (! isGuaranteed)
The comments should remain though :)

Reply:
would also avoid the speculative wording as it could cause further confusion. I believe preemptable should be fine in this context